src/plugins/ucx/ucx_backend.cpp

```diff
@@ -1244,3 +1286,92 @@ nixl_status_t nixlUcxEngine::genNotif(const std::string &remote_agent, const std
     }
     return NIXL_SUCCESS;
 }
+
+/****************************************
+ * Thread to worker mapping
+*****************************************/
+
+void nixlUcxEngine::threadMapDestructor(void *arg)
+{
+    nixlUcxWorker *worker = static_cast<nixlUcxWorker *>(arg);
+
+    /* When a thread exits, disassociate it from the worker,
+     * so that the worker could be reused by another thread.
+     */
+    nixlUcxEngine* engine = static_cast<nixlUcxEngine*>(worker->getEngine());
+    engine->pushFreeWorker(worker);
+}
+
+nixl_status_t nixlUcxEngine::initThreadMapping()
+{
+    int err = pthread_key_create(&keyThreadToWorker, threadMapDestructor);
+    if (err) {
+        NIXL_ERROR << "Failed to create keyThreadToWorker, err:" << err;
+        return NIXL_ERR_BACKEND;
+    }
+    return NIXL_SUCCESS;
+}
+
+nixl_status_t nixlUcxEngine::destroyThreadMapping()
+{
+    int err = pthread_key_delete(keyThreadToWorker);
+    if (err) {
+        NIXL_WARN << "Failed to delete keyThreadToWorker, err:" << err;
+        return NIXL_ERR_INVALID_PARAM;
+    }
+    return NIXL_SUCCESS;
+}
+
+nixlUcxWorker *nixlUcxEngine::getFreeDedicatedWorker() const
+{
+    if (freeWorkers.size() == 0) {
+        return nullptr;
+    }
+
+    const std::lock_guard<std::mutex> lock(workersMutex);
+    nixlUcxWorker *worker = freeWorkers.front();
+    freeWorkers.pop();
+    return worker;
```

Comment:
Better use empty().
Do we really need this micro-optimization? Maybe:

```suggestion
    const std::lock_guard<std::mutex> lock(workersMutex);
    if (freeWorkers.empty()) {
        return nullptr;
    }

    nixlUcxWorker *worker = freeWorkers.front();
    freeWorkers.pop();
    return worker;
```

Reply:
Thanks for catching this. Yes, should check empty after lock